File: packages/less/src/less/tree/attribute.test.js
import { describe, it, expect, vi } from 'vitest';
import Attribute from './attribute';

describe('Attribute', () => {
    it('should create an attribute with key, operator, and value', () => {
        const attr = new Attribute('data-test', '=', 'value');
        expect(attr.key).toBe('data-test');
        expect(attr.op).toBe('=');
        expect(attr.value).toBe('value');
        expect(attr.cif).toBeUndefined();
    });

    it('should create an attribute with key, operator, value, and cif', () => {
        const attr = new Attribute('data-test', '=', 'value', 'if (condition)');
        expect(attr.key).toBe('data-test');
        expect(attr.op).toBe('=');
        expect(attr.value).toBe('value');
        expect(attr.cif).toBe('if (condition)');
    });

    it('should evaluate key and value when they are evaluable', () => {
        const context = {
            variables: { key: 'data-test', value: 'test-value' }
        };
        const evaluableKey = { eval: (ctx) => ctx.variables.key };
        const evaluableValue = { eval: (ctx) => ctx.variables.value };

        const attr = new Attribute(evaluableKey, '=', evaluableValue);
        const evaluated = attr.eval(context);

        expect(evaluated.key).toBe('data-test');
        expect(evaluated.value).toBe('test-value');
    });

    it('should not evaluate non-evaluable key and value', () => {
        const context = { variables: {} };
        const attr = new Attribute('data-test', '=', 'value');
        const evaluated = attr.eval(context);

        expect(evaluated.key).toBe('data-test');
        expect(evaluated.value).toBe('value');
    });

    it('should generate CSS with key only when no operator is present', () => {
        const attr = new Attribute('data-test');
        expect(attr.toCSS()).toBe('[data-test]');
    });

    it('should generate CSS with key and value when operator is present', () => {
        const attr = new Attribute('data-test', '=', 'value');
        expect(attr.toCSS()).toBe('[data-test=value]');
    });

    it('should generate CSS with key, value, and cif when present', () => {
        const attr = new Attribute('data-test', '=', 'value', 'if (condition)');
        expect(attr.toCSS()).toBe('[data-test=value if (condition)]');
    });

    it('should handle evaluable key and value in toCSS', () => {
        const context = {
            variables: { key: 'data-test', value: 'test-value' }
        };
        const evaluableKey = {
            eval: (ctx) => ctx.variables.key,
            toCSS: (ctx) => ctx.variables.key
        };
        const evaluableValue = {
            eval: (ctx) => ctx.variables.value,
            toCSS: (ctx) => ctx.variables.value
        };

        const attr = new Attribute(evaluableKey, '=', evaluableValue);
        expect(attr.toCSS(context)).toBe('[data-test=test-value]');
    });

    it('should add CSS to output in genCSS', () => {
        const attr = new Attribute('data-test', '=', 'value');
        const output = { add: vi.fn() };
        attr.genCSS({}, output);
        expect(output.add).toHaveBeenCalledWith('[data-test=value]');
    });

    it('should handle empty string values', () => {
        const attr = new Attribute('data-test', '=', '');
        expect(attr.toCSS()).toBe('[data-test=]');
    });

    it('should handle special characters in key and value', () => {
        const attr = new Attribute('data-test[0]', '=', 'value with spaces');
        expect(attr.toCSS()).toBe('[data-test[0]=value with spaces]');
    });

    it('should handle different operators', () => {
        const operators = ['=', '~=', '|=', '^=', '$=', '*='];
        operators.forEach((op) => {
            const attr = new Attribute('data-test', op, 'value');
            expect(attr.toCSS()).toBe(`[data-test${op}value]`);
        });
    });

    it('should handle evaluable key with toCSS but without eval', () => {
        const context = { variables: { key: 'data-test' } };
        const evaluableKey = { toCSS: (ctx) => ctx.variables.key };
        const attr = new Attribute(evaluableKey, '=', 'value');
        expect(attr.toCSS(context)).toBe('[data-test=value]');
    });

    it('should handle evaluable value with toCSS but without eval', () => {
        const context = { variables: { value: 'test-value' } };
        const evaluableValue = { toCSS: (ctx) => ctx.variables.value };
        const attr = new Attribute('data-test', '=', evaluableValue);
        expect(attr.toCSS(context)).toBe('[data-test=test-value]');
    });

    it('should preserve operator in eval when key and value are not evaluable', () => {
        const context = { variables: {} };
        const attr = new Attribute('data-test', '~=', 'value');
        const evaluated = attr.eval(context);
        expect(evaluated.op).toBe('~=');
    });

    it('should handle empty context in eval', () => {
        const attr = new Attribute('data-test', '=', 'value');
        const evaluated = attr.eval();
        expect(evaluated.key).toBe('data-test');
        expect(evaluated.value).toBe('value');
    });

    it('should handle undefined context in toCSS', () => {
        const attr = new Attribute('data-test', '=', 'value');
        expect(attr.toCSS(undefined)).toBe('[data-test=value]');
    });

    it('should handle null context in toCSS', () => {
        const attr = new Attribute('data-test', '=', 'value');
        expect(attr.toCSS(null)).toBe('[data-test=value]');
    });

    it('should handle null/undefined key', () => {
        const attr = new Attribute(null, '=', 'value');
        expect(() => attr.toCSS()).toThrow(
            "Cannot read properties of null (reading 'toCSS')"
        );

        const attr2 = new Attribute(undefined, '=', 'value');
        expect(() => attr2.toCSS()).toThrow(
            "Cannot read properties of undefined (reading 'toCSS')"
        );
    });

    it('should handle null/undefined value', () => {
        const attr = new Attribute('data-test', '=', null);
        expect(() => attr.toCSS()).toThrow(
            "Cannot read properties of null (reading 'toCSS')"
        );

        const attr2 = new Attribute('data-test', '=', undefined);
        expect(() => attr2.toCSS()).toThrow(
            "Cannot read properties of undefined (reading 'toCSS')"
        );
    });

    it('should handle null/undefined operator', () => {
        const attr = new Attribute('data-test', null, 'value');
        expect(attr.toCSS()).toBe('[data-test]');

        const attr2 = new Attribute('data-test', undefined, 'value');
        expect(attr2.toCSS()).toBe('[data-test]');
    });

    it('should handle evaluable object with only eval method', () => {
        const context = { variables: { key: 'data-test' } };
        const evaluableKey = { eval: (ctx) => ctx.variables.key };
        const attr = new Attribute(evaluableKey, '=', 'value');
        const evaluated = attr.eval(context);
        expect(evaluated.key).toBe('data-test');
        expect(evaluated.toCSS()).toBe('[data-test=value]');
    });

    it('should handle evaluable object with only toCSS method', () => {
        const context = { variables: { key: 'data-test' } };
        const evaluableKey = { toCSS: (ctx) => ctx.variables.key };
        const attr = new Attribute(evaluableKey, '=', 'value');
        expect(attr.toCSS(context)).toBe('[data-test=value]');
    });

    it('should handle special characters in cif condition', () => {
        const attr = new Attribute(
            'data-test',
            '=',
            'value',
            'if (condition && value > 0)'
        );
        expect(attr.toCSS()).toBe(
            '[data-test=value if (condition && value > 0)]'
        );
    });

    it('should handle whitespace in various positions', () => {
        const attr = new Attribute(' data-test ', '=', ' value ');
        expect(attr.toCSS()).toBe('[ data-test = value ]');
    });

    it('should handle very long key/value combinations', () => {
        const longKey = 'data-' + 'x'.repeat(1000);
        const longValue = 'value-' + 'y'.repeat(1000);
        const attr = new Attribute(longKey, '=', longValue);
        expect(attr.toCSS()).toBe(`[${longKey}=${longValue}]`);
    });

    it('should handle malformed evaluable objects gracefully', () => {
        const malformedKey = { eval: null };
        const attr = new Attribute(malformedKey, '=', 'value');
        expect(attr.toCSS()).toBe('[[object Object]=value]');
    });

    it('should handle invalid operator values', () => {
        const attr = new Attribute('data-test', 'invalid', 'value');
        expect(attr.toCSS()).toBe('[data-testinvalidvalue]');
    });
});

File: packages/less/src/less/go_parser/attribute_test.go
package go_parser

import (
	"fmt"
	"strings"
	"testing"
)

// MockEvaluable implements Evaluable interface for testing
type MockEvaluable struct {
	value any
}

func (m *MockEvaluable) Eval(context any) any {
	if ctx, ok := context.(map[string]any); ok {
		if val, ok := ctx["variables"].(map[string]any); ok {
			return val[m.value.(string)]
		}
	}
	return m.value
}

// MockCSSable implements CSSable interface for testing
type MockCSSable struct {
	value any
}

func (m *MockCSSable) ToCSS(context any) string {
	if ctx, ok := context.(map[string]any); ok {
		if val, ok := ctx["variables"].(map[string]any); ok {
			return val[m.value.(string)].(string)
		}
	}
	return fmt.Sprintf("%v", m.value)
}

func TestAttribute(t *testing.T) {
	tests := []struct {
		name     string
		attr     *Attribute
		context  any
		expected string
	}{
		{
			name:     "basic attribute with key only",
			attr:     NewAttribute("data-test", "", nil, ""),
			expected: "[data-test]",
		},
		{
			name:     "attribute with key and value",
			attr:     NewAttribute("data-test", "=", "value", ""),
			expected: "[data-test=value]",
		},
		{
			name:     "attribute with key, value, and cif",
			attr:     NewAttribute("data-test", "=", "value", "if (condition)"),
			expected: "[data-test=value if (condition)]",
		},
		{
			name: "attribute with evaluable key and value",
			attr: NewAttribute(
				&MockEvaluable{value: "key"},
				"=",
				&MockEvaluable{value: "value"},
				"",
			),
			context: map[string]any{
				"variables": map[string]any{
					"key":   "data-test",
					"value": "test-value",
				},
			},
			expected: "[data-test=test-value]",
		},
		{
			name: "attribute with cssable key and value",
			attr: NewAttribute(
				&MockCSSable{value: "key"},
				"=",
				&MockCSSable{value: "value"},
				"",
			),
			context: map[string]any{
				"variables": map[string]any{
					"key":   "data-test",
					"value": "test-value",
				},
			},
			expected: "[data-test=test-value]",
		},
		{
			name:     "attribute with empty string value",
			attr:     NewAttribute("data-test", "=", "", ""),
			expected: "[data-test=]",
		},
		{
			name:     "attribute with special characters",
			attr:     NewAttribute("data-test[0]", "=", "value with spaces", ""),
			expected: "[data-test[0]=value with spaces]",
		},
		{
			name:     "attribute with different operators",
			attr:     NewAttribute("data-test", "~=", "value", ""),
			expected: "[data-test~=value]",
		},
		{
			name:     "attribute with nil operator",
			attr:     NewAttribute("data-test", "", "value", ""),
			expected: "[data-test]",
		},
		{
			name:     "attribute with empty cif",
			attr:     NewAttribute("data-test", "=", "value", ""),
			expected: "[data-test=value]",
		},
		{
			name:     "attribute with special characters in cif",
			attr:     NewAttribute("data-test", "=", "value", "if (condition && value > 0)"),
			expected: "[data-test=value if (condition && value > 0)]",
		},
		{
			name:     "attribute with whitespace",
			attr:     NewAttribute(" data-test ", "=", " value ", ""),
			expected: "[ data-test = value ]",
		},
		{
			name:     "attribute with very long key/value",
			attr:     NewAttribute("data-"+strings.Repeat("x", 1000), "=", "value-"+strings.Repeat("y", 1000), ""),
			expected: fmt.Sprintf("[data-%s=value-%s]", strings.Repeat("x", 1000), strings.Repeat("y", 1000)),
		},
		{
			name:     "attribute with all operators",
			attr:     NewAttribute("data-test", "*=", "value", ""),
			expected: "[data-test*=value]",
		},
		{
			name:     "attribute with nil key",
			attr:     NewAttribute(nil, "=", "value", ""),
			expected: "[]",
		},
		{
			name:     "attribute with nil value",
			attr:     NewAttribute("data-test", "=", nil, ""),
			expected: "[data-test=]",
		},
		{
			name:     "attribute with evaluable object with only eval method",
			attr:     NewAttribute(&MockEvaluable{value: "key"}, "=", "value", ""),
			context: map[string]any{
				"variables": map[string]any{
					"key": "data-test",
				},
			},
			expected: "[data-test=value]",
		},
		{
			name:     "attribute with evaluable object with only toCSS method",
			attr:     NewAttribute(&MockCSSable{value: "key"}, "=", "value", ""),
			context: map[string]any{
				"variables": map[string]any{
					"key": "data-test",
				},
			},
			expected: "[data-test=value]",
		},
		{
			name:     "attribute with malformed evaluable object",
			attr:     NewAttribute(&struct{}{}, "=", "value", ""),
			expected: "[&{}=value]",
		},
		{
			name:     "attribute with invalid operator",
			attr:     NewAttribute("data-test", "invalid", "value", ""),
			expected: "[data-testinvalidvalue]",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.context != nil {
				evaluated := tt.attr.Eval(tt.context)
				if evaluated.ToCSS(tt.context) != tt.expected {
					t.Errorf("Eval().ToCSS() = %v, want %v", evaluated.ToCSS(tt.context), tt.expected)
				}
			} else {
				if tt.attr.ToCSS(nil) != tt.expected {
					t.Errorf("ToCSS() = %v, want %v", tt.attr.ToCSS(nil), tt.expected)
				}
			}
		})
	}
}

func TestAttributeGenCSS(t *testing.T) {
	attr := NewAttribute("data-test", "=", "value", "")
	output := &CSSOutput{
		Add: func(chunk any, fileInfo any, index any) {
			if chunk != "[data-test=value]" {
				t.Errorf("GenCSS() output = %v, want %v", chunk, "[data-test=value]")
			}
		},
		IsEmpty: func() bool { return false },
	}
	attr.GenCSS(nil, output)
} 